src/attr.rs

```diff
@@ -578,9 +597,30 @@ pub mod parsing {
     }
 
     impl Parse for Meta {
+        /// Attempts to parse tokens as a `Meta`.
+        ///
+        /// Historically this assumed the first token encountered is an `Ident`.
+        /// However, derive attributes such as `#[derive(serde::Serialize)]`
+        /// have been valid since Rust 1.30:
+        ///
+        /// <https://github.com/rust-lang/rust/blob/stable/RELEASES.md#version-1300-2018-10-25>
+        ///
+        /// Therefore, we need to accept `Path`s as well, while erring on
+        /// invalid usages of `Path`s.
+        ///
+        /// For backward compatibility, we will treat single segment paths as
+        /// `Meta::Word`s.
         fn parse(input: ParseStream) -> Result<Self> {
-            let ident = input.call(Ident::parse_any)?;
-            parse_meta_after_ident(ident, input)
+            // We assume `Path` values to contain `::` as single words are parsed as `Meta::Word`
+            // for backward compatibility.
+            if input.peek(Token![::]) || input.peek2(Token![::]) {
+                let path = input.call(Path::parse_mod_style)?;
+                Ok(Meta::Path(path))
```

Comment:
What about `#[path::to = ""]` and `#[path::to("")]`?

Reply:
hmm, earlier intentionally decided to leave those out. If we include that, there's two umbrella designs that are apparent:

* Expand the variants of `Meta` to have both the `Ident` and `Path` versions:

    ```rust
    ListIdent(MetaListIdent {
        pub ident: Ident,
        pub meta_list: MetaList,
    }),
    ListPath(MetaListPath {
        pub path: Path,
        pub meta_list: MetaList,
    })

    // and
    MetaList {
        pub paren_token: token::Paren,
        pub nested: Punctuated<NestedMeta, Token![,]>,
    }
    ```

* Or, replace `Ident` with `Path`

    ```rust
    List(MetaList {
        pub path: Path,
        pub paren_token: token::Paren,
        pub nested: Punctuated<NestedMeta, Token![,]>,
    })
    ```

Or something in between. The second option isn't exactly straightforward -- right now `Ident::parse_any` happily parses keywords, so substituting `Ident` with `Path` would mean figuring out:

* Should `Path` have a parse that accepts keywords?
* If yes, it should *likely* be a separate parse function -- we don't want to parse keywords everywhere.
* If no, then we are back to choosing one of those overall designs (or hybrid).

Is there a preference?

Reply:
I would like to keep Meta as an enum with three variants but use Path instead of Ident in each. The parsing for Meta should accept keywords in the path but the parsing can be implemented in attr.rs without exposing a new public parser function on Path.